File: docs/beginner/module-05-data-sources-indexers/code-samples/javascript/03_cosmos_db_indexer.js
/**
 * Azure Cosmos DB Indexer Example
 * 
 * This example demonstrates how to create and manage indexers for Azure Cosmos DB data sources.
 * It covers JSON document processing, change feed integration, and partition key handling.
 */

const { SearchIndexClient, SearchIndexerClient, SearchClient } = require('@azure/search-documents');
const { AzureKeyCredential } = require('@azure/core-auth');
require('dotenv').config();

// Configuration
const SEARCH_ENDPOINT = process.env.SEARCH_ENDPOINT;
const SEARCH_API_KEY = process.env.SEARCH_API_KEY;
const COSMOS_CONNECTION_STRING = process.env.COSMOS_CONNECTION_STRING;

// Resource names
const DATA_SOURCE_NAME = 'cosmos-hotels-datasource';
const INDEX_NAME = 'hotels-cosmos-index';
const INDEXER_NAME = 'hotels-cosmos-indexer';
const CONTAINER_NAME = 'hotels';

class CosmosDbIndexerExample {
    constructor() {
        this.validateConfiguration();
        
        const credential = new AzureKeyCredential(SEARCH_API_KEY);
        this.indexClient = new SearchIndexClient(SEARCH_ENDPOINT, credential);
        this.indexerClient = new SearchIndexerClient(SEARCH_ENDPOINT, credential);
    }

    validateConfiguration() {
        if (!SEARCH_ENDPOINT || !SEARCH_API_KEY || !COSMOS_CONNECTION_STRING) {
            throw new Error('Missing required environment variables. Check your .env file.');
        }
        
        console.log('✅ Configuration validated');
        console.log(`📍 Search Endpoint: ${SEARCH_ENDPOINT}`);
        console.log(`🗃️ Data Source: ${DATA_SOURCE_NAME}`);
        console.log(`📊 Index: ${INDEX_NAME}`);
        console.log(`⚙️ Indexer: ${INDEXER_NAME}`);
    }

    async createCosmosDataSource() {
        console.log('\n🔗 Creating Cosmos DB data source...');

        const dataSource = {
            name: DATA_SOURCE_NAME,
            type: 'cosmosdb',
            connectionString: COSMOS_CONNECTION_STRING,
            container: {
                name: CONTAINER_NAME,
                query: 'SELECT * FROM c WHERE c._ts >= @HighWaterMark ORDER BY c._ts'
            },
            dataChangeDetectionPolicy: {
                '@odata.type': '#Microsoft.Azure.Search.HighWaterMarkChangeDetectionPolicy',
                highWaterMarkColumnName: '_ts'
            },
            description: 'Hotel data from Cosmos DB with change feed detection'
        };

        try {
            const result = await this.indexerClient.createOrUpdateDataSourceConnection(dataSource);
            console.log(`✅ Data source '${DATA_SOURCE_NAME}' created successfully`);
            console.log(`   Type: ${result.type}`);
            console.log(`   Container: ${result.container.name}`);
            console.log(`   Change Detection: ${result.dataChangeDetectionPolicy['@odata.type']}`);
            console.log(`   High Water Mark: ${result.dataChangeDetectionPolicy.highWaterMarkColumnName}`);
            return result;
        } catch (error) {
            console.log(`❌ Error creating data source: ${error.message}`);
            throw error;
        }
    }

    async createHotelsIndex() {
        console.log('\n📊 Creating hotels index...');

        const fields = [
            { name: 'id', type: 'Edm.String', key: true, searchable: false },
            { name: 'hotelName', type: 'Edm.String', searchable: true, sortable: true },
            { name: 'description', type: 'Edm.String', searchable: true, analyzer: 'en.lucene' },
            { name: 'category', type: 'Edm.String', filterable: true, facetable: true },
            { name: 'rating', type: 'Edm.Double', filterable: true, sortable: true, facetable: true },
            {
                name: 'address',
                type: 'Edm.ComplexType',
                fields: [
                    { name: 'street', type: 'Edm.String', searchable: true },
                    { name: 'city', type: 'Edm.String', searchable: true, filterable: true, facetable: true },
                    { name: 'state', type: 'Edm.String', searchable: true, filterable: true, facetable: true },
                    { name: 'zipCode', type: 'Edm.String', filterable: true },
                    { name: 'country', type: 'Edm.String', filterable: true, facetable: true }
                ]
            },
            { name: 'amenities', type: 'Collection(Edm.String)', searchable: true, facetable: true },
            { name: 'lastRenovated', type: 'Edm.DateTimeOffset', filterable: true, sortable: true },
            { name: '_ts', type: 'Edm.Int64', filterable: true, sortable: true },
            { name: 'location', type: 'Edm.GeographyPoint', filterable: true, sortable: true }
        ];

        const index = { name: INDEX_NAME, fields };

        try {
            const result = await this.indexClient.createOrUpdateIndex(index);
            console.log(`✅ Index '${INDEX_NAME}' created successfully`);
            console.log(`   Total Fields: ${result.fields.length}`);
            
            console.log('   Key fields for hotel data:');
            const keyFields = ['id', 'hotelName', 'category', 'rating', 'address'];
            result.fields.filter(f => keyFields.includes(f.name)).forEach(field => {
                const attributes = [];
                if (field.key) attributes.push('key');
                if (field.searchable) attributes.push('searchable');
                if (field.filterable) attributes.push('filterable');
                if (field.facetable) attributes.push('facetable');
                
                const fieldType = field.type === 'Edm.ComplexType' ? 'Complex' : field.type;
                console.log(`     - ${field.name} (${fieldType}) [${attributes.join(', ')}]`);
            });
            
            return result;
        } catch (error) {
            console.log(`❌ Error creating index: ${error.message}`);
            throw error;
        }
    }

    async createCosmosIndexer() {
        console.log('\n⚙️ Creating Cosmos DB indexer...');

        const fieldMappings = [
            { sourceFieldName: 'id', targetFieldName: 'id' },
            { sourceFieldName: 'hotelName', targetFieldName: 'hotelName' },
            { sourceFieldName: 'description', targetFieldName: 'description' },
            { sourceFieldName: 'category', targetFieldName: 'category' },
            { sourceFieldName: 'rating', targetFieldName: 'rating' },
            { sourceFieldName: 'address', targetFieldName: 'address' },
            { sourceFieldName: 'amenities', targetFieldName: 'amenities' },
            { sourceFieldName: 'lastRenovated', targetFieldName: 'lastRenovated' },
            { sourceFieldName: '_ts', targetFieldName: '_ts' }
        ];

        const indexer = {
            name: INDEXER_NAME,
            dataSourceName: DATA_SOURCE_NAME,
            targetIndexName: INDEX_NAME,
            fieldMappings,
            description: 'Indexer for hotel data from Cosmos DB',
            parameters: {
                batchSize: 100,
                maxFailedItems: 10,
                maxFailedItemsPerBatch: 5,
                configuration: { parsingMode: 'json' }
            }
        };

        try {
            const result = await this.indexerClient.createOrUpdateIndexer(indexer);
            console.log(`✅ Indexer '${INDEXER_NAME}' created successfully`);
            console.log(`   Data Source: ${result.dataSourceName}`);
            console.log(`   Target Index: ${result.targetIndexName}`);
            console.log(`   Field Mappings: ${result.fieldMappings?.length || 0}`);
            console.log(`   Batch Size: ${result.parameters?.batchSize || 0}`);
            return result;
        } catch (error) {
            console.log(`❌ Error creating indexer: ${error.message}`);
            throw error;
        }
    }

    async runAndMonitorIndexer() {
        console.log(`\n🚀 Starting Cosmos DB indexer: ${INDEXER_NAME}`);

        try {
            await this.indexerClient.runIndexer(INDEXER_NAME);
            console.log('✅ Indexer started successfully');

            const startTime = Date.now();
            const maxWaitTime = 5 * 60 * 1000;

            while (Date.now() - startTime < maxWaitTime) {
                const status = await this.indexerClient.getIndexerStatus(INDEXER_NAME);
                const currentTime = new Date().toLocaleTimeString();
                console.log(`\n⏰ ${currentTime} - Status: ${status.status}`);

                if (status.lastResult) {
                    const result = status.lastResult;
                    console.log(`   🏨 Hotels processed: ${result.itemCount || 0}`);
                    console.log(`   ❌ Hotels failed: ${result.failedItemCount || 0}`);

                    if (result.itemCount && result.itemCount > 0 && result.endTime && result.startTime) {
                        const duration = (new Date(result.endTime) - new Date(result.startTime)) / 1000;
                        if (duration > 0) {
                            const rate = result.itemCount / duration;
                            console.log(`   📊 Processing rate: ${rate.toFixed(2)} hotels/sec`);
                        }
                    }

                    if (result.errors && result.errors.length > 0) {
                        console.log(`   ⚠️ Recent errors:`);
                        result.errors.slice(0, 3).forEach(error => {
                            console.log(`     - ${error.errorMessage}`);
                        });
                    }

                    if (result.warnings && result.warnings.length > 0) {
                        console.log(`   ⚠️ Warnings: ${result.warnings.length}`);
                    }
                }

                if (status.status === 'success' || status.status === 'error') {
                    console.log(`\n🎉 Indexer execution completed with status: ${status.status}`);
                    break;
                }

                await this.sleep(10000);
            }
        } catch (error) {
            console.log(`❌ Error running indexer: ${error.message}`);
            throw error;
        }
    }

    async testHotelSearch() {
        console.log('\n🔍 Testing hotel search...');

        const searchClient = new SearchClient(SEARCH_ENDPOINT, INDEX_NAME, new AzureKeyCredential(SEARCH_API_KEY));

        const testQueries = [
            { name: 'All hotels', searchText: '*', filter: null, orderBy: null, facets: null },
            { name: 'Luxury hotels', searchText: 'luxury', filter: null, orderBy: null, facets: null },
            { name: 'High-rated hotels', searchText: '*', filter: 'rating ge 4.0', orderBy: ['rating desc'], facets: null },
            { name: 'Hotels by city', searchText: '*', filter: null, orderBy: null, facets: ['address/city'] },
            { name: 'Hotels with amenities', searchText: '*', filter: "amenities/any(a: a eq 'WiFi')", orderBy: null, facets: null }
        ];

        for (const test of testQueries) {
            console.log(`\n   🔍 ${test.name}:`);
            try {
                const searchOptions = {
                    top: test.facets ? 0 : 3,
                    select: ['hotelName', 'category', 'rating', 'address']
                };

                if (test.filter) searchOptions.filter = test.filter;
                if (test.orderBy) searchOptions.orderBy = test.orderBy;
                if (test.facets) searchOptions.facets = test.facets;

                const results = await searchClient.search(test.searchText, searchOptions);

                if (test.facets) {
                    if (results.facets && results.facets['address/city']) {
                        console.log(`      Cities found:`);
                        results.facets['address/city'].slice(0, 5).forEach(facet => {
                            console.log(`        - ${facet.value}: ${facet.count} hotels`);
                        });
                    }
                } else {
                    const documents = [];
                    for await (const result of results.results) {
                        documents.push(result);
                    }

                    console.log(`      Found ${documents.length} results`);

                    documents.slice(0, 2).forEach((result, index) => {
                        const hotelName = result.document.hotelName || 'N/A';
                        const category = result.document.category || 'N/A';
                        const rating = result.document.rating || 0;
                        const city = result.document.address?.city || 'N/A';
                        console.log(`      ${index + 1}. ${hotelName} (${category}) - ${rating}⭐ in ${city}`);
                    });
                }
            } catch (error) {
                console.log(`      ❌ Error: ${error.message}`);
            }
        }
    }

    demonstrateChangeFeed() {
        console.log('\n🔄 Change Feed Detection:');
        console.log('   Cosmos DB indexer uses the \'_ts\' field for change detection');
        console.log('   Benefits:');
        console.log('   - Automatic detection of new and updated documents');
        console.log('   - Efficient incremental updates');
        console.log('   - Built-in ordering by timestamp');
        console.log('   - No additional configuration required');
        console.log('\n   The indexer query includes: WHERE c._ts >= @HighWaterMark ORDER BY c._ts');
        console.log('   This ensures only changed documents are processed on subsequent runs');
    }

    showCleanupOptions() {
        console.log('\n🧹 Cleanup options:');
        console.log('   To clean up resources, call:');
        console.log(`   - await indexerClient.deleteIndexer('${INDEXER_NAME}');`);
        console.log(`   - await indexClient.deleteIndex('${INDEX_NAME}');`);
        console.log(`   - await indexerClient.deleteDataSourceConnection('${DATA_SOURCE_NAME}');`);
    }

    async cleanup() {
        try {
            await this.indexerClient.deleteIndexer(INDEXER_NAME);
            await this.indexClient.deleteIndex(INDEX_NAME);
            await this.indexerClient.deleteDataSourceConnection(DATA_SOURCE_NAME);
            console.log('✅ Resources cleaned up successfully');
        } catch (error) {
            console.log(`⚠️ Cleanup warning: ${error.message}`);
        }
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async run() {
        console.log('🚀 Azure Cosmos DB Indexer Example');
        console.log('='.repeat(50));

        try {
            const dataSource = await this.createCosmosDataSource();
            const index = await this.createHotelsIndex();
            const indexer = await this.createCosmosIndexer();

            await this.runAndMonitorIndexer();
            await this.testHotelSearch();
            this.demonstrateChangeFeed();
            this.showCleanupOptions();

            console.log('\n✅ Cosmos DB indexer example completed successfully!');
            console.log('\nKey takeaways:');
            console.log('- Cosmos DB change feed provides efficient incremental updates');
            console.log('- Complex fields handle nested JSON structures well');
            console.log('- Collection fields are perfect for arrays like amenities');
            console.log('- Faceted search works great with categorical hotel data');
            console.log('- Higher batch sizes work well with JSON documents');
        } catch (error) {
            console.log(`\n❌ Example failed: ${error.message}`);
            throw error;
        }
    }
}

async function main() {
    const example = new CosmosDbIndexerExample();
    try {
        await example.run();
    } catch (error) {
        console.error('Application failed:', error.message);
        process.exit(1);
    }
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = CosmosDbIndexerExample;

File: docs/beginner/module-05-data-sources-indexers/code-samples/python/03_cosmos_db_indexer.py
#!/usr/bin/env python3
"""
Azure Cosmos DB Indexer Example

This script demonstrates how to create and manage indexers for Azure Cosmos DB data sources.
It covers JSON document processing, change feed integration, and partition key handling.

Prerequisites:
- Azure AI Search service
- Azure Cosmos DB account with sample data
- Admin API key or managed identity
- Required Python packages installed
"""

import os
import time
from datetime import datetime
from dotenv import load_dotenv

from azure.search.documents.indexes import SearchIndexClient, SearchIndexerClient
from azure.search.documents.indexes.models import (
    SearchIndex, SearchField, SearchFieldDataType, SearchIndexer,
    SearchIndexerDataContainer, SearchIndexerDataSourceConnection,
    SearchIndexerDataSourceType, HighWaterMarkChangeDetectionPolicy,
    FieldMapping, SimpleField, SearchableField, ComplexField
)
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError

# Load environment variables
load_dotenv()

# Configuration
SEARCH_ENDPOINT = os.getenv('SEARCH_ENDPOINT')
SEARCH_API_KEY = os.getenv('SEARCH_API_KEY')
COSMOS_CONNECTION_STRING = os.getenv('COSMOS_CONNECTION_STRING')

# Resource names
DATA_SOURCE_NAME = "cosmos-hotels-datasource"
INDEX_NAME = "hotels-cosmos-index"
INDEXER_NAME = "hotels-cosmos-indexer"
CONTAINER_NAME = "hotels"

def validate_configuration():
    """Validate that all required configuration is present."""
    required_vars = [SEARCH_ENDPOINT, SEARCH_API_KEY, COSMOS_CONNECTION_STRING]
    if not all(required_vars):
        raise ValueError("Missing required environment variables. Check your .env file.")
    
    print("✅ Configuration validated")
    print(f"📍 Search Endpoint: {SEARCH_ENDPOINT}")
    print(f"🗃️ Data Source: {DATA_SOURCE_NAME}")
    print(f"📊 Index: {INDEX_NAME}")
    print(f"⚙️ Indexer: {INDEXER_NAME}")

def create_cosmos_data_source(indexer_client):
    """Create a data source connection to Azure Cosmos DB."""
    print("\n🔗 Creating Cosmos DB data source...")
    
    data_source = SearchIndexerDataSourceConnection(
        name=DATA_SOURCE_NAME,
        type=SearchIndexerDataSourceType.COSMOS_DB,
        connection_string=COSMOS_CONNECTION_STRING,
        container=SearchIndexerDataContainer(
            name=CONTAINER_NAME,
            query="SELECT * FROM c WHERE c._ts >= @HighWaterMark ORDER BY c._ts"
        ),
        data_change_detection_policy=HighWaterMarkChangeDetectionPolicy(
            high_water_mark_column_name="_ts"
        ),
        description="Hotel data from Cosmos DB with change feed detection"
    )
    
    try:
        result = indexer_client.create_or_update_data_source_connection(data_source)
        print(f"✅ Data source '{DATA_SOURCE_NAME}' created successfully")
        
        # Display configuration
        print(f"   Type: {result.type}")
        print(f"   Container: {result.container.name}")
        print(f"   Change Detection: {type(result.data_change_detection_policy).__name__}")
        print(f"   High Water Mark: {result.data_change_detection_policy.high_water_mark_column_name}")
        
        return result
    except HttpResponseError as e:
        print(f"❌ Error creating data source: {e.message}")
        raise

def create_hotels_index(index_client):
    """Create a search index optimized for hotel data from Cosmos DB."""
    print("\n📊 Creating hotels index...")
    
    # Define index fields for hotel data
    fields = [
        SimpleField(name="id", type=SearchFieldDataType.String, key=True),
        SearchableField(name="hotelName", type=SearchFieldDataType.String, sortable=True),
        SearchableField(name="description", type=SearchFieldDataType.String, analyzer_name="en.lucene"),
        SimpleField(name="category", type=SearchFieldDataType.String, filterable=True, facetable=True),
        SimpleField(name="rating", type=SearchFieldDataType.Double, filterable=True, sortable=True, facetable=True),
        
        # Complex field for address
        ComplexField(name="address", fields=[
            SearchableField(name="street", type=SearchFieldDataType.String),
            SearchableField(name="city", type=SearchFieldDataType.String, filterable=True, facetable=True),
            SearchableField(name="state", type=SearchFieldDataType.String, filterable=True, facetable=True),
            SimpleField(name="zipCode", type=SearchFieldDataType.String, filterable=True),
            SimpleField(name="country", type=SearchFieldDataType.String, filterable=True, facetable=True)
        ]),
        
        # Collection field for amenities
        SearchableField(name="amenities", type=SearchFieldDataType.Collection(SearchFieldDataType.String), facetable=True),
        
        # Date fields
        SimpleField(name="lastRenovated", type=SearchFieldDataType.DateTimeOffset, filterable=True, sortable=True),
        SimpleField(name="_ts", type=SearchFieldDataType.Int64, filterable=True, sortable=True),
        
        # Location field for geo-search (if coordinates available)
        SimpleField(name="location", type=SearchFieldDataType.GeographyPoint, filterable=True, sortable=True)
    ]
    
    index = SearchIndex(
        name=INDEX_NAME,
        fields=fields
    )
    
    try:
        result = index_client.create_or_update_index(index)
        print(f"✅ Index '{INDEX_NAME}' created successfully")
        print(f"   Total Fields: {len(result.fields)}")
        
        # Display key fields
        print("   Key fields for hotel data:")
        key_fields = ['id', 'hotelName', 'category', 'rating', 'address']
        for field in result.fields:
            if field.name in key_fields:
                attributes = []
                if hasattr(field, 'key') and field.key:
                    attributes.append('key')
                if hasattr(field, 'searchable') and field.searchable:
                    attributes.append('searchable')
                if hasattr(field, 'filterable') and field.filterable:
                    attributes.append('filterable')
                if hasattr(field, 'facetable') and field.facetable:
                    attributes.append('facetable')
                
                field_type = field.type if hasattr(field, 'type') else 'Complex'
                print(f"     - {field.name} ({field_type}) [{', '.join(attributes)}]")
        
        return result
    except HttpResponseError as e:
        print(f"❌ Error creating index: {e.message}")
        raise

def create_cosmos_indexer(indexer_client):
    """Create an indexer to process hotel data from Cosmos DB."""
    print("\n⚙️ Creating Cosmos DB indexer...")
    
    # Field mappings for Cosmos DB data
    field_mappings = [
        FieldMapping(source_field_name="id", target_field_name="id"),
        FieldMapping(source_field_name="hotelName", target_field_name="hotelName"),
        FieldMapping(source_field_name="description", target_field_name="description"),
        FieldMapping(source_field_name="category", target_field_name="category"),
        FieldMapping(source_field_name="rating", target_field_name="rating"),
        FieldMapping(source_field_name="address", target_field_name="address"),
        FieldMapping(source_field_name="amenities", target_field_name="amenities"),
        FieldMapping(source_field_name="lastRenovated", target_field_name="lastRenovated"),
        FieldMapping(source_field_name="_ts", target_field_name="_ts")
    ]
    
    indexer = SearchIndexer(
        name=INDEXER_NAME,
        data_source_name=DATA_SOURCE_NAME,
        target_index_name=INDEX_NAME,
        field_mappings=field_mappings,
        description="Indexer for hotel data from Cosmos DB",
        parameters={
            "batchSize": 100,  # Larger batch for JSON documents
            "maxFailedItems": 10,
            "maxFailedItemsPerBatch": 5,
            "configuration": {
                "parsingMode": "json"
            }
        }
    )
    
    try:
        result = indexer_client.create_or_update_indexer(indexer)
        print(f"✅ Indexer '{INDEXER_NAME}' created successfully")
        print(f"   Data Source: {result.data_source_name}")
        print(f"   Target Index: {result.target_index_name}")
        print(f"   Field Mappings: {len(result.field_mappings)}")
        print(f"   Batch Size: {result.parameters.get('batchSize', 'default')}")
        
        return result
    except HttpResponseError as e:
        print(f"❌ Error creating indexer: {e.message}")
        raise

def run_and_monitor_indexer(indexer_client):
    """Run the indexer and monitor hotel data processing."""
    print(f"\n🚀 Starting Cosmos DB indexer: {INDEXER_NAME}")
    
    try:
        # Start the indexer
        indexer_client.run_indexer(INDEXER_NAME)
        print("✅ Indexer started successfully")
        
        # Monitor execution
        start_time = time.time()
        max_wait_time = 300  # 5 minutes
        
        while time.time() - start_time < max_wait_time:
            status = indexer_client.get_indexer_status(INDEXER_NAME)
            
            current_time = datetime.now().strftime('%H:%M:%S')
            print(f"\n⏰ {current_time} - Status: {status.status}")
            
            if status.last_result:
                result = status.last_result
                print(f"   🏨 Hotels processed: {result.item_count or 0}")
                print(f"   ❌ Hotels failed: {result.failed_item_count or 0}")
                
                # Show processing rate if available
                if result.item_count and result.item_count > 0 and result.end_time and result.start_time:
                    duration = (result.end_time - result.start_time).total_seconds()
                    if duration > 0:
                        rate = result.item_count / duration
                        print(f"   📊 Processing rate: {rate:.2f} hotels/sec")
                
                # Show any errors
                if result.errors:
                    print(f"   ⚠️ Recent errors:")
                    for error in result.errors[:3]:  # Show first 3 errors
                        print(f"     - {error.error_message}")
                
                # Show warnings if any
                if result.warnings:
                    print(f"   ⚠️ Warnings: {len(result.warnings)}")
            
            if status.status in ["success", "error"]:
                print(f"\n🎉 Indexer execution completed with status: {status.status}")
                break
            
            time.sleep(10)
        else:
            print(f"\n⏰ Monitoring timeout reached ({max_wait_time}s)")
        
        return status
        
    except HttpResponseError as e:
        print(f"❌ Error running indexer: {e.message}")
        raise

def test_hotel_search():
    """Test search functionality on the indexed hotel data."""
    print("\n🔍 Testing hotel search...")
    
    from azure.search.documents import SearchClient
    
    credential = AzureKeyCredential(SEARCH_API_KEY)
    search_client = SearchClient(SEARCH_ENDPOINT, INDEX_NAME, credential)
    
    test_queries = [
        {
            'name': 'All hotels',
            'query': {'search_text': '*', 'top': 5}
        },
        {
            'name': 'Luxury hotels',
            'query': {'search_text': 'luxury', 'top': 3}
        },
        {
            'name': 'High-rated hotels',
            'query': {
                'search_text': '*', 
                'filter': "rating ge 4.0", 
                'order_by': ['rating desc'],
                'top': 3
            }
        },
        {
            'name': 'Hotels by city',
            'query': {
                'search_text': '*', 
                'facets': ['address/city'],
                'top': 0  # Just get facets
            }
        },
        {
            'name': 'Hotels with amenities',
            'query': {
                'search_text': '*',
                'filter': "amenities/any(a: a eq 'WiFi')",
                'top': 3
            }
        }
    ]
    
    for test in test_queries:
        print(f"\n   🔍 {test['name']}:")
        try:
            results = search_client.search(**test['query'])
            
            if test['name'] == 'Hotels by city':
                # Handle facet results
                facets = results.get_facets()
                if facets and 'address/city' in facets:
                    print(f"      Cities found:")
                    for facet in facets['address/city'][:5]:
                        print(f"        - {facet['value']}: {facet['count']} hotels")
            else:
                # Handle regular search results
                results_list = list(results)
                print(f"      Found {len(results_list)} results")
                
                for i, result in enumerate(results_list[:2]):  # Show first 2 results
                    hotel_name = result.get('hotelName', 'N/A')
                    category = result.get('category', 'N/A')
                    rating = result.get('rating', 0)
                    city = result.get('address', {}).get('city', 'N/A') if result.get('address') else 'N/A'
                    print(f"      {i+1}. {hotel_name} ({category}) - {rating}⭐ in {city}")
                
        except Exception as e:
            print(f"      ❌ Error: {str(e)}")

def demonstrate_change_feed():
    """Demonstrate how change feed detection works with Cosmos DB."""
    print("\n🔄 Change Feed Detection:")
    print("   Cosmos DB indexer uses the '_ts' field for change detection")
    print("   Benefits:")
    print("   - Automatic detection of new and updated documents")
    print("   - Efficient incremental updates")
    print("   - Built-in ordering by timestamp")
    print("   - No additional configuration required")
    print("\n   The indexer query includes: WHERE c._ts >= @HighWaterMark ORDER BY c._ts")
    print("   This ensures only changed documents are processed on subsequent runs")

def cleanup_resources(index_client, indexer_client):
    """Clean up created resources (optional)."""
    print("\n🧹 Cleanup options:")
    print("   To clean up resources, run:")
    print(f"   - Delete indexer: indexer_client.delete_indexer('{INDEXER_NAME}')")
    print(f"   - Delete index: index_client.delete_index('{INDEX_NAME}')")
    print(f"   - Delete data source: indexer_client.delete_data_source_connection('{DATA_SOURCE_NAME}')")

def main():
    """Main execution function."""
    print("🚀 Azure Cosmos DB Indexer Example")
    print("=" * 50)
    
    try:
        # Validate configuration
        validate_configuration()
        
        # Initialize clients
        credential = AzureKeyCredential(SEARCH_API_KEY)
        index_client = SearchIndexClient(SEARCH_ENDPOINT, credential)
        indexer_client = SearchIndexerClient(SEARCH_ENDPOINT, credential)
        
        # Create resources
        data_source = create_cosmos_data_source(indexer_client)
        index = create_hotels_index(index_client)
        indexer = create_cosmos_indexer(indexer_client)
        
        # Run and monitor indexer
        status = run_and_monitor_indexer(indexer_client)
        
        # Test search functionality
        if status.status == "success":
            test_hotel_search()
        
        # Demonstrate change feed concepts
        demonstrate_change_feed()
        
        # Show cleanup options
        cleanup_resources(index_client, indexer_client)
        
        print("\n✅ Cosmos DB indexer example completed successfully!")
        print("\nKey takeaways:")
        print("- Cosmos DB change feed provides efficient incremental updates")
        print("- Complex fields handle nested JSON structures well")
        print("- Collection fields are perfect for arrays like amenities")
        print("- Faceted search works great with categorical hotel data")
        print("- Higher batch sizes work well with JSON documents")
        
    except Exception as e:
        print(f"\n❌ Example failed: {str(e)}")
        raise

if __name__ == "__main__":
    main()